CryptoAnalysis/src/main/java/crypto/ProviderDetection.java

```diff
@@ -0,0 +1,187 @@
+									
+								if((ruleFound) && (methodName.matches("getInstance")) && (parameterCount==2) ) {
+									Value vl = exp.getArg(1);
+									String strType = getProviderType(vl);
+										
+									if(strType.matches("java.security.Provider")) {
+										rules = chooseRulesWhenProviderTypeProvider(stmt, method, vl, icfg, refName, rules);
+									}
+										
+									else if (strType.matches("java.lang.String")) {
+										rules = chooseRulesWhenProviderTypeString(vl, refName, rules);
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+		
+		return rules;
+	}
+	
+	//methods used from the "doAnalysis() method"
+	//-----------------------------------------------------------------------------------------------------------------
+	
+	private String getProviderType(Value vl) {
+		Type type = vl.getType();
+		String strType = type.toString();
+		return strType;
+	}
+	
+	private List<CryptSLRule> chooseRulesWhenProviderTypeString(Value vl, String refName, List<CryptSLRule> rules) {
+		String provider = vl.toString();
+		
+		File rule = new File(".\\.\\.\\test\\resources\\"+provider+"\\"+refName+".cryptslbin");
+		if(rule.exists()) {
+			//delete the default rules and load the new rules from the "Provider" directory
+			String newDirectory = ".\\.\\.\\test\\resources\\"+provider;
+			
+			rules.clear();
+			
+			File[] listFiles = new File(newDirectory).listFiles();
+			for (File file : listFiles) {
+				if (file != null && file.getName().endsWith(".cryptslbin")) {
+					rules.add(CryptSLRuleReader.readFromFile(file));
+				}
+			}
+		}
+		return rules;
+	}
+	
+	private List<CryptSLRule> chooseRulesWhenProviderTypeProvider(JAssignStmt stmt, SootMethod method, Value vl, JimpleBasedInterproceduralCFG icfg, String refName, List<CryptSLRule> rules) {
+		String provider;
+		
+		BackwardQuery query = new BackwardQuery(new Statement(stmt,method), new Val(vl, method));
+		
+		//Create a Boomerang solver.
+		Boomerang solver = new Boomerang(new DefaultBoomerangOptions(){
+			public boolean onTheFlyCallGraph() {
+				//Must be turned of if no SeedFactory is specified.
+				return false;
+			};
+		}) {
+			@Override
+			public BiDiInterproceduralCFG<Unit, SootMethod> icfg() {
+				return icfg;
+			}
+
+			@Override
+			public boomerang.seedfactory.SeedFactory<NoWeight> getSeedFactory() {
+				return null;
+			}
+		};
+		
+		//Submit query to the solver.
+		BackwardBoomerangResults<NoWeight> backwardQueryResults = solver.solve(query);
+		solver.debugOutput();
+		
+		Map<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> map = backwardQueryResults.getAllocationSites();
+		Entry<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> entry = map.entrySet().iterator().next();
+		ForwardQuery forwardQuery = entry.getKey();
+		
+		Val forwardQueryVal = forwardQuery.var();
+		Value value = forwardQueryVal.value();
+		Type valueType = value.getType();
+		String valueTypeString = valueType.toString();
```

Comment:
This method and chooseProviderWhenProviderTypeString are virtually the same, starting here. Rather call it here and return its return value. Makes maintenance easier.

Reply:
That is true. I will make the corresponding changes. Thanks